README.markdown
# remote_table

Open Google Docs spreadsheets, local or remote XLSX, XLS, ODS, CSV (comma separated), TSV (tab separated), other delimited, fixed-width files.

Tested on MRI 1.8, MRI 1.9, and JRuby 1.6.7+. Thread-safe.

## Sponsor

<p><a href="http://faraday.io"><img src="http://cdn2.hubspot.net/hubfs/515497/img/logo.svg" alt="Faraday logo"/></a></p>

We use [`remote_table`](https://github.com/seamusabshere/remote_table) for [data-driven marketing at Faraday](http://faraday.io).

## Example

    >> require 'remote_table'
    => true
    >> t = RemoteTable.new 'http://www.fueleconomy.gov/FEG/epadata/98guide6.zip', :filename => '98guide6.csv'
    => #<RemoteTable:0x00000101b87390 @download_count_mutex=#<Mutex:0x00000101b87228>, @extend_bang_mutex=#<Mutex:0x00000101b871d8>, @errata_mutex=#<Mutex:0x00000101b871b0>, @cache=[], @download_count=0, @url="http://www.fueleconomy.gov/FEG/epadata/98guide6.zip", @format=nil, @headers=:first_row, @compression=:zip, @packing=nil, @streaming=false, @warn_on_multiple_downloads=true, @delimiter=",", @sheet=nil, @keep_blank_rows=false, @form_data=nil, @skip=0, @internal_encoding="UTF-8", @row_xpath=nil, @column_xpath=nil, @row_css=nil, @column_css=nil, @glob=nil, @filename="98guide6.csv", @transform_settings=nil, @cut=nil, @crop=nil, @schema=nil, @schema_name=nil, @pre_select=nil, @pre_reject=nil, @errata_settings=nil, @other_options={}, @transformer=#<RemoteTable::Transformer:0x00000101b8c2f0 @t=#<RemoteTable:0x00000101b87390 ...>, @legacy_transformer_mutex=#<Mutex:0x00000101b8c2a0>>, @local_copy=#<RemoteTable::LocalCopy:0x00000101b8bf58 @t=#<RemoteTable:0x00000101b87390 ...>, @encoded_io_mutex=#<Mutex:0x00000101b8be18>, @generate_mutex=#<Mutex:0x00000101b8bdc8>>>
    >> t.rows.length
    => 806
    >> t.rows.first.length
    => 26
    >> require 'pp'
    => true
    >> pp t[23]
    {"Class"=>"TWO SEATERS",
     "Manufacturer"=>"PORSCHE",
     "carline name"=>"BOXSTER",
     "displ"=>"2.5",
     "cyl"=>"6",
     "trans"=>"Manual(M5)",
     "drv"=>"R",
     "cty"=>"19",
     "hwy"=>"26",
     "cmb"=>"22",
     "ucty"=>"21.2",
     "uhwy"=>"33.9499",
     "ucmb"=>"25.5114",
     "fl"=>"P",
     "G"=>"",
     "T"=>"",
     "S"=>"",
     "2pv"=>"",
     "2lv"=>"",
     "4pv"=>"",
     "4lv"=>"",
     "hpv"=>"",
     "hlv"=>"",
     "fcost"=>"956",
     "eng dscr"=>"",
     "trans dscr"=>""}

## Columns and rows

* If there are headers, you get an <code>Array</code> of <code>Hash</code>es with **string keys**.
* If you set <code>:headers => false</code>, then you get an <code>Array</code> of <code>Array</code>s.

## Row keys

Row keys are **strings**. Row keys are NOT symbolized.

    row['foobar'] # correct
    row[:foobar]  # incorrect

You can call <code>symbolize_keys</code> yourself, but we don't do it automatically to avoid creating loads of garbage symbols.

## Supported formats

<table>
  <tr>
    <th>Format</th>
    <th>Notes</th>
    <th>Library</th>
  </tr>
  <tr>
    <td>Delimited (CSV, TSV, etc.)</td>
    <td>All <code>RemoteTable::Delimited::PASSTHROUGH_CSV_SETTINGS</code>, for example <code>:col_sep</code>, are passed directly to fastercsv.</td>
    <td>
      <a href="http://fastercsv.rubyforge.org/">fastercsv</a> (1.8);
      <a href="http://www.ruby-doc.org/stdlib-1.9.3/libdoc/csv/rdoc/index.html">stdlib</code></a> (1.9)
    </td>
  </tr>
  <tr>
    <td>Fixed width</td>
    <td>You have to set up a <code>:schema</code>.</td>
    <td><a href="https://github.com/seamusabshere/fixed_width">fixed_width-multibyte</a></td>
  </tr>
  <tr>
    <td>HTML</td>
    <td>See XML.</td>
    <td><a href="http://nokogiri.org/">nokogiri</a></td>
  </tr>
  <tr>
    <td>ODS</td>
    <td></td>
    <td><a href="http://roo.rubyforge.org/">roo</a></td>
  </tr>
  <tr>
    <td>XLS</td>
    <td></td>
    <td><a href="http://roo.rubyforge.org/">roo</a></td>
  </tr>
  <tr>
    <td>XLSX</td>
    <td></td>
    <td><a href="http://roo.rubyforge.org/">roo</a></td>
  </tr>
  <tr>
    <td>XML</td>
    <td>The idea is to set up a <code>:row_[xpath|css]</code> and (optionally) a <code>:column_[xpath|css]</code>.</td>
    <td><a href="http://nokogiri.org/">nokogiri</a></td>
  </tr>
  <tr>
    <td>JSON</td>
    <td>Force JSON format using <code>format: :json</code> and define root nodes using <code>root_node: 'data'</code></td>
    <td><a href="http://www.ruby-doc.org/stdlib-2.0.0/libdoc/json/rdoc/JSON.html">JSON</a></td>
  </tr>
</table>

## Compression and packing

You can directly pick a file out of a remote archive using <code>:filename</code> or use a <code>:glob</code>.

* zip
* tar
* bz2
* gz
* exe (treated as zip)

## Encoding

Everything is forced into UTF-8. You can improve the quality of the conversion by specifying the original encoding with `:encoding`

* ASCII-8BIT and BINARY are equal
* ISO-8859-1 and Latin1 are equal

## More examples

    RemoteTable.new('https://spreadsheets.google.com/pub?key=0AoQJbWqPrREqdHRNaVpSUWw2Z2VhN3RUV25yYWdQX2c&output=csv')

    # aircraft fuel use equations derived from EMEP/EEA and ICAO
    RemoteTable.new('https://docs.google.com/spreadsheet/pub?key=0AoQJbWqPrREqdEhYenF3dGt1T0Y1cTdneUNsNjV0dEE&output=csv')

    # distance classes from the WRI business travel tool and UK DEFRA/DECC GHG Conversion Factors for Company Reporting
    RemoteTable.new('https://spreadsheets.google.com/pub?key=0AoQJbWqPrREqdFBKM0xWaUhKVkxDRmdBVkE3VklxY2c&hl=en&gid=0&output=csv')

    # seat classes used in the WRI GHG Protocol calculation tools
    RemoteTable.new('https://docs.google.com/spreadsheet/pub?key=0AoQJbWqPrREqdG9EdmxybG1wdC1iU3JRYXNkMGhvSnc&output=csv')

    # pure automobile fuels
    RemoteTable.new('https://spreadsheets.google.com/pub?key=0AoQJbWqPrREqdE9xTEdueFM2R0diNTgxUlk1QXFSb2c&gid=0&output=csv')

    # blended automobile fuels
    RemoteTable.new('https://spreadsheets.google.com/pub?key=0AoQJbWqPrREqdEswNGIxM0U4U0N1UUppdWw2ejJEX0E&gid=0&output=csv')

    # A list of hybrid make model years derived from the EPA fuel economy guide
    RemoteTable.new('https://docs.google.com/spreadsheet/pub?hl=en_US&hl=en_US&key=0AoQJbWqPrREqdGtzekE4cGNoRGVmdmZMaTNvOWluSnc&output=csv')

    # BTS aircraft type lookup table
    RemoteTable.new("http://www.transtats.bts.gov/Download_Lookup.asp?Lookup=L_AIRCRAFT_TYPE",
                    :errata => { :url => RemoteTable.new('https://spreadsheets.google.com/spreadsheet/pub?key=0AoQJbWqPrREqdEZ2d3JQMzV5T1o1T3JmVlFyNUZxdEE&output=csv' })

    # aircraft made by whitelisted manufacturers whose ICAO code starts with 'B' from the FAA
    # for definition of `Aircraft::Guru` and `manufacturer_whitelist?` see https://github.com/brighterplanet/earth/blob/master/lib/earth/air/aircraft/data_miner.rb
    RemoteTable.new("http://www.faa.gov/air_traffic/publications/atpubs/CNT/5-2-B.htm",
                    :encoding => 'windows-1252',
                    :row_xpath => '//table/tr[2]/td/table/tr',
                    :column_xpath => 'td',
                    :errata => { :url => 'https://spreadsheets.google.com/spreadsheet/pub?key=0AoQJbWqPrREqdGVBRnhkRGhSaVptSDJ5bXJGbkpUSWc&output=csv', :responder => Aircraft::Guru.new },
                    :select => proc { |record| manufacturer_whitelist? record['Manufacturer'] })

    # OpenFlights.org airports database
    RemoteTable.new('https://openflights.svn.sourceforge.net/svnroot/openflights/openflights/data/airports.dat',
                    :headers => %w{ id name city country_name iata_code icao_code latitude longitude altitude timezone daylight_savings },
                    :select => proc { |record| record['iata_code'].present? },
                    :errata => { :url => RemoteTable.new('https://spreadsheets.google.com/pub?key=0AoQJbWqPrREqdFc2UzhQYU5PWEQ0N21yWFZGNmc2a3c&gid=0&output=csv', :responder => Airport::Guru.new }) # see https://github.com/brighterplanet/earth/blob/master/lib/earth/air/aircraft/data_miner.rb

    # T100 flight segment data for #{month.strftime('%B %Y')}
    # for definition of `form_data` and `FlightSegment::Guru` see https://github.com/brighterplanet/earth/blob/master/lib/earth/air/flight_segment/data_miner.rb
    RemoteTable.new('http://www.transtats.bts.gov/DownLoad_Table.asp',
                    :form_data => form_data,
                    :compression => :zip,
                    :glob => '/*.csv',
                    :errata => { :url => RemoteTable.new('https://spreadsheets.google.com/spreadsheet/pub?key=0AoQJbWqPrREqdGxpYU1qWFR3d0syTVMyQVVOaDd0V3c&output=csv', :responder => FlightSegment::Guru.new },
                    :select => proc { |record| record['DEPARTURES_PERFORMED'].to_i > 0 })

    # 1995 Fuel Economy Guide
    # for definition of `:fuel_economy_guide_b` and `AutomobileMakeModelYearVariant::ParserB` see https://github.com/brighterplanet/earth/blob/master/lib/earth/automobile/automobile_make_model_year_variant/data_miner.rb
    RemoteTable.new("http://www.fueleconomy.gov/FEG/epadata/95mfgui.zip",
                    :filename => '95MFGUI.DAT',
                    :format => :fixed_width,
                    :cut => '13-',
                    :schema_name => :fuel_economy_guide_b,
                    :select => proc { |row| row['model'].present? and (row['suppress_code'].blank? or row['suppress_code'].to_f == 0) and row['state_code'] == 'F' },
                    :transform => { :class => AutomobileMakeModelYearVariant::ParserB, :year => 1995 },
                    :errata => { :url => "https://docs.google.com/spreadsheet/pub?key=0AoQJbWqPrREqdDkxTElWRVlvUXB3Uy04SDhSYWkzakE&output=csv", :responder => AutomobileMakeModelYearVariant::Guru.new })

    # 1998 Fuel Economy Guide
    # for definition of `AutomobileMakeModelYearVariant::ParserC` see https://github.com/brighterplanet/earth/blob/master/lib/earth/automobile/automobile_make_model_year_variant/data_miner.rb
    RemoteTable.new('http://www.fueleconomy.gov/FEG/epadata/98guide6.zip',
                    :filename => '98guide6.csv',

CHANGELOG
3.3.3 / 2016-02-01

* Enhancements

  * New option :stop_after_untitled_headers

* Bug fixes

  * Fix Excel support via roo-xls
  * Treat headers: true as :first_row in roo

3.3.2 / 2015-10-01

* Bug fixes

  * Correctly require ActiveSupport

3.3.1 / 2014-08-01

* Enhancements

  * Support for JSON endpoints support added to the library

3.3.0 / 2014-01-14

* Breaking changes

  * switch to safer HashDigest.digest3 for #row_hash - BREAKING CHANGE if you use row_hash as a unique id!

3.2.1 / 2014-01-07

* Bug fixes

  * Don't blow up when requiring activesupport (something about "uninitialized constant ActiveSupport::Deprecation::MethodWrapper")

3.2.0 / 2014-01-06

* Breaking changes

  * switch to faster HashDigest.digest2 for #row_hash - BREAKING CHANGE if you use row_hash as a unique id!

3.1.0 / 2014-01-06

* Breaking changes

  * Stop requiring rubyzip2 on behalf of Roo per @rossmeissl https://github.com/seamusabshere/remote_table/issues/15 ... you may have to require it yourself

* Enhancements

  * by default, treat single and double quotes as "not special" in pipe- or tab-delimited files - thanks @dwillis https://github.com/seamusabshere/remote_table/pull/14

3.0.0 / 2013-11-26

* Breaking changes

  * Include columns with blank headers as "untitled_N"
  * skip blank lines at the top of CSV files when looking for headers [rc2]
  * require rubyzip2 for Roo compatibility with Ruby 1.9 [rc2]
  * :parser object must respond to #call(row), not #parse(row) [rc1]
  * Whitespace now automatically "normalized" (/\s+/ -> ' ', stripped) for CSV [rc1]
  * Strip whitespace from headers unless you provide them as an array [beta]
  * Not passing anything options on to CSV... use :delimiter instead of :col_sep [beta]
  * Just use :encoding to specify a file's internal/initial/original encoding. No more :internal_encoding, this was jargony. [alpha]
  * No more shapefile support - hard to imagine a shapefile as a table [alpha]
  * No more :transform, legacy or not [alpha]
  * For :errata, now you pass an object that responds to #rejects?(row) and #correct!(row) [alpha]
  * No more :errata_settings [alpha]

* Enhancements

  * eagerly load format-specific methods (instead of waiting for #each to be called) [rc2]
  * use new Roo class names (Roo::OpenOffice instead of Roo::Openoffice) [rc2]
  * Support relative paths with spaces [beta]
  * :parser option takes an object that responds to #parse(row) and returns an array of one or more rows. [alpha]
  * RemoteTable.transpose(url, key_key, value_key) helper [alpha]

* Bug fixes

  * pass CSV options to header parser - before, options like delimiter weren't passed so it couldn't read TSVs [rc2]

3.0.0.rc2 / 2013-08-30

* Breaking changes

  * skip blank lines at the top of CSV files when looking for headers
  * require rubyzip2 for Roo compatibility with Ruby 1.9

* Enhancements

  * eagerly load format-specific methods (instead of waiting for #each to be called)
  * use new Roo class names (Roo::OpenOffice instead of Roo::Openoffice)

* Bug fixes

  * pass CSV options to header parser - before, options like delimiter weren't passed so it couldn't read TSVs

3.0.0.rc1 / 2013-08-08

* Breaking changes

  * :parser object must respond to #call(row), not #parse(row)
  * Whitespace now automatically "normalized" (/\s+/ -> ' ', stripped) for CSV

3.0.0.beta / 2013-07-30

* Breaking changes

  * Strip whitespace from headers unless you provide them as an array
  * Not passing anything options on to CSV... use :delimiter instead of :col_sep
  * Include columns with blank headers as "empty_N"

* Enhancements

  * Support relative paths with spaces

3.0.0.alpha / 2013-07-25

* Breaking changes

  * Just use :encoding to specify a file's internal/initial/original encoding. No more :internal_encoding, this was jargony.
  * No more shapefile support - hard to imagine a shapefile as a table
  * No more :transform, legacy or not
  * For :errata, now you pass an object that responds to #rejects?(row) and #correct!(row)
  * No more :errata_settings

* Enhancements

  * :parser option takes an object that responds to #parse(row) and returns an array of one or more rows.
  * RemoteTable.transpose(url, key_key, value_key) helper

2.1.2 / 2013-10-08

* Bug fixes

  * Correctly detect format from filename - don't get confused if a filename has "xls" in it (thanks @activefx #10)

2.1.1 / 2013-03-25

* Enhancements

  * Work with new roo - thanks @towerhe !

2.1.0 / 2012-10-01

* Enhancements

  * Allow import of ESRI shapefiles

2.0.2 / 2012-06-20

* Secrets revealed

  * Indeed it is OK to omit the file:// as long as you are using an absolute path to a local file... e.g., RemoteTable.new('/path/to/foo')

* Bug fixes

  * Don't bork UTF-8 by incorrectly treating \xAD as a soft hyphen

2.0.1 / 2012-05-16

* Enhancements

  * For single-string transliteration to UTF-8, use ActiveSupport::Inflector.transliterate instead of Iconv - possibly a future-proof solution.

lib/remote_table.rb
if ::RUBY_VERSION < '1.9' and $KCODE != 'UTF8'
  ::Kernel.warn "[remote_table] Ruby 1.8 detected, setting $KCODE to UTF8 so that ActiveSupport::Multibyte works properly."
  $KCODE = 'UTF8'
end

require 'thread'
require 'hash_digest'
require 'active_support'
require 'active_support/core_ext'

require 'remote_table/local_copy'

require 'remote_table/plaintext'
require 'remote_table/processed_by_roo'
require 'remote_table/processed_by_nokogiri'
require 'remote_table/xls'
require 'remote_table/xlsx'
require 'remote_table/delimited'
require 'remote_table/ods'
require 'remote_table/fixed_width'
require 'remote_table/html'
require 'remote_table/xml'
require 'remote_table/yaml'
require 'remote_table/json'

class Hash
  # Added by remote_table to store a hash (think checksum) of the data with which a particular Hash is initialized.
  # @return [String]
  attr_accessor :row_hash
end

class Array
  # Added by remote_table to store a hash (think checksum) of the data with which a particular Array is initialized.
  # @return [String]
  attr_accessor :row_hash
end

# Open Google Docs spreadsheets, local or remote XLSX, XLS, ODS, CSV (comma separated), TSV (tab separated), other delimited, fixed-width files.
class RemoteTable
  class << self
    # Transpose two columns into a mapping from one to the other.
    def transpose(url, key_key, value_key, options = {})
      new(url, options).inject({}) do |memo, row|
        memo[row[key_key]] = row[value_key]
        memo
      end
    end

    # Guess compression based on URL. Used internally.
    # @return [Symbol,nil]
    def guess_compression(url)
      extname = extname(url).downcase
      case extname
      when /gz/, /gunzip/
        :gz
      when /zip/
        :zip
      when /bz2/, /bunzip2/
        :bz2
      when /exe/
        :exe
      end
    end

    # Guess packing from URL. Used internally.
    # @return [Symbol,nil]
    def guess_packing(url)
      basename = basename(url).downcase
      if basename.include?('.tar') or basename.include?('.tgz')
        :tar
      end
    end

    # Guess file format from the basename. Since a file might be decompressed and/or pulled out of an archive with a glob, this usually can't be called until a file is downloaded.
    # @return [Symbol,nil]
    def guess_format(basename)
      case basename.to_s.downcase.strip
      when /ods\z/, /open_?office\z/
        :ods
      when /xlsx\z/, /excelx\z/
        :xlsx
      when /xls\z/, /excel\z/
        :xls
      when /csv\z/, /tsv\z/, /delimited\z/
        # note that there is no RemoteTable::Csv class - it's normalized to :delimited
        :delimited
      when /fixed_?width\z/
        :fixed_width
      when /html?\z/
        :html
      when /xml\z/
        :xml
      when /yaml\z/, /yml\z/
        :yaml
      when /json\z/
        :json
      end
    end

    # Given a Google Docs spreadsheet URL, make sure it uses CSV output.
    # @return [String]
    def google_spreadsheet_csv_url(url)
      uri = ::URI.parse url
      params = uri.query.split('&')
      params.delete_if { |param| param.start_with?('output=') }
      params << 'output=csv'
      uri.query = params.join('&')
      uri.to_s
    end

    def normalize_whitespace(v)
      v = v.to_s.dup
      v.gsub! WHITESPACE, SINGLE_SPACE
      v.strip!
      v
    end

    private

    def basename(url)
      ::File.basename path(url)
    end

    def extname(url)
      ::File.extname path(url)
    end

    def path(url)
      if url.include?('://')
        ::URI.parse(url).path
      else
        File.expand_path url
      end
    end
  end

  WHITESPACE = /\s+/
  SINGLE_SPACE = ' '
  EXTERNAL_ENCODING = 'UTF-8'
  EXTERNAL_ENCODING_ICONV = 'UTF-8//TRANSLIT'
  GOOGLE_DOCS_SPREADSHEET = [
    /docs.google.com/i,
    /spreadsheets.google.com/i
  ]
  VALID = {
    :compression => [:gz, :zip, :bz2, :exe],
    :packing => [:tar],
    :format => [:xlsx, :xls, :delimited, :ods, :fixed_width, :html, :xml, :yaml, :csv, :json],
  }
  DEFAULT = {
    :streaming => false,
    :warn_on_multiple_downloads => true,
    :headers => :first_row,
    :keep_blank_rows => false,
    :skip => 0,
    :encoding => 'UTF-8',
    :stop_after_untitled_headers => false,
  }
  OLD_SETTING_NAMES = {
    :pre_select => [:select],
    :pre_reject => [:reject],
    :delimiter  => [:col_sep],
  }

  include ::Enumerable

  # The URL of the local or remote file.
  #
  # @example Local
  #   file:///Users/myuser/Desktop/holidays.csv
  #
  # @example Local using an absolute path
  #   /Users/myuser/Desktop/holidays.csv
  #
  # @example Remote
  #   http://data.brighterplanet.com/countries.csv
  #
  # @return [String]
  attr_reader :url

  # @private
  # A cache of rows, created unless +:streaming+ is enabled.
  # @return [Array<Hash,Array>]
  attr_reader :cache

  # @private
  # How many times this file has been downloaded. RemoteTable will emit a warning if you download it more than once.
  # @return [Integer]
  attr_reader :download_count

  # @private
  # Used internally to access to a downloaded copy of the file.
  # @return [RemoteTable::LocalCopy]
  attr_reader :local_copy

  # Whether to stream the rows without caching them. Saves memory, but you have to re-download the file every time you enumerate its rows. Defaults to false.
  # @return [true,false]
  attr_reader :streaming

  # Whether to warn the user on multiple downloads. Defaults to true.
  # @return [true,false]
  attr_reader :warn_on_multiple_downloads

  # Headers specified by the user: +:first_row+ (the default), +false+, or a list of headers.
  # @return [:first_row,false,Array<String>]
  attr_reader :headers

  # Quote character for delimited files.
  #
  # Defaults to double quotes.
  #
  # @return [String]
  attr_reader :quote_char

  # The sheet specified by the user as a number or a string.
  # @return[String,Integer]
  attr_reader :sheet

  # Whether to keep blank rows. Default is false.
  # @return [true,false]
  attr_reader :keep_blank_rows

  # Form data to POST in the download request. It should probably be in +application/x-www-form-urlencoded+.
  # @return [String]
  attr_reader :form_data

  # How many rows to skip at the beginning of the file or table. Default is 0.
  # @return [Integer]
  attr_reader :skip

  # The original encoding of the source file. Default is UTF-8.
  # @return [String]
  attr_reader :encoding

  # The delimiter, a.k.a. column separator. Passed to Ruby CSV as +:col_sep+. Default is ','.
  # @return [String]
  attr_reader :delimiter

  # The XPath used to find rows in HTML or XML.
  # @return [String]
  attr_reader :row_xpath

  # The XPath used to find columns in HTML or XML.
  # @return [String]
  attr_reader :column_xpath

  # The CSS selector used to find rows in HTML or XML.
  # @return [String]
  attr_reader :row_css

  # The CSS selector used to find columns in HTML or XML.
  # @return [String]
  attr_reader :column_css

  # The format of the source file. Can be +:xlsx+, +:xls+, +:delimited+, +:ods+, +:fixed_width+, +:html+, +:xml+, +:yaml+, +:json+.
  # @return [Symbol]
  attr_reader :format

  # The compression type. Guessed from URL if not provided. +:gz+, +:zip+, +:bz2+, and +:exe+ (treated as +:zip+) are supported.
  # @return [Symbol]
  attr_reader :compression

  # The packing type. Guessed from URL if not provided. Only +:tar+ is supported.
  # @return [Symbol]
  attr_reader :packing

  # The glob used to pick a file out of an archive.
  #
  # @return [String]
  #
  # @example Pick out the only CSV in a ZIP file
  #   RemoteTable.new 'http://www.fueleconomy.gov/FEG/epadata/08data.zip', :glob => '/*.csv'
  attr_reader :glob

  # The filename, which can be used to pick a file out of an archive.
  #
  # @return [String]
  #
  # @example Specify the filename to get out of a ZIP file
  #   RemoteTable.new 'http://www.fueleconomy.gov/FEG/epadata/08data.zip', :filename => '2008_FE_guide_ALL_rel_dates_-no sales-for DOE-5-1-08.csv'
  attr_reader :filename

  # Pick specific columns out of a plaintext file using an argument to the UNIX [+cut+ utility](http://en.wikipedia.org/wiki/Cut_%28Unix%29).
  #
  # @return [String]
  #
  # @example Pick ALMOST out of ABCDEFGHIJKLMNOPQRSTUVWXYZ
  #   # $ echo ABCDEFGHIJKLMNOPQRSTUVWXYZ | cut -c '1,12,13,15,19,20'
  #   # ALMOST
  #   RemoteTable.new 'file:///atoz.txt', :cut => '1,12,13,15,19,20'
  attr_reader :cut

  # Use a range of rows in a plaintext file.
  #
  # @return [Range]
  #
  # @example Only take rows 21 through 37
  #   RemoteTable.new("http://www.eia.gov/emeu/cbecs/cbecs2003/detailed_tables_2003/2003set10/2003excel/C17.xls",
  #                   :headers => false,
  #                   :select => proc { |row| CbecsEnergyIntensity::NAICS_CODE_SYNTHESIZER.call(row) },
  #                   :crop => (21..37))
  attr_reader :crop

  # The fixed-width schema, given as a multi-dimensional array.
  #
  # @return [Array<Array{String,Integer,Hash}>]
  #
  # @example From the tests
  #   RemoteTable.new('http://cloud.github.com/downloads/seamusabshere/remote_table/test2.fixed_width.txt',
  #                    :format => :fixed_width,
  #                    :skip => 1,
  #                    :schema => [[ 'header4', 10, { :type => :string }  ],
  #                                [  'spacer',  1 ],
  #                                [  'header5', 10, { :type => :string } ],
  #                                [  'spacer',  12 ],
  #                                [  'header6', 10, { :type => :string } ]])
  attr_reader :schema

  # If you somehow already defined a fixed-width schema (so you can re-use it?), specify it here.
  # @return [String,Symbol]
  attr_reader :schema_name

  # A proc that decides whether to include a row. Previously passed as +:select+.
  # @return [Proc]
  attr_reader :pre_select

  # A proc that decides whether to include a row. Previously passed as +:reject+.
  # @return [Proc]
  attr_reader :pre_reject

  # An object that responds to #rejects?(row) and #correct!(row). Applied after creating +row_hash+.
  #
  # * #rejects?(row) - if row should be treated like it doesn't exist
  # * #correct!(row) - destructively update a row to fix something
  #
  # See the Errata library at https://github.com/seamusabshere/errata for an example implementation.
  #
  # @return [Hash]
  attr_reader :errata

  # The format of the source file. Can be specified as: :xlsx, :xls, :delimited (aka :csv), :ods, :fixed_width, :html, :xml, :yaml :json
  #
  # Note: treats all +docs.google.com+ and +spreadsheets.google.com+ URLs as +:delimited+.
  #
  # Default: guessed from file extension (which is usually the same as the URL, but sometimes not if you pick out a specific file from an archive)
  #
  # @return [Hash]
  attr_reader :format

  # The root node of the json document. Specified as a string.
  #
  # Default: nil; no root node.
  #
  # @return [String]
  attr_reader :root_node

  # When to trim untitled headers. Set this to 100 to prevent more than 100 untitled headers being created; the rest will be silently discarded.
  #
  # Note: This is effectively a right trim... the counting starts from the left.
  #
  # Default: false, don't try
  #
  # @return [Integer]
  attr_reader :stop_after_untitled_headers

  # @private
  class NullParser
    def call(row)
      [row]
    end
  end

  # An object that responds to #call(row) and returns an array of one or more rows.
  #
  # @return [#call]
  def parser
    @final_parser ||= (@parser || NullParser.new)
  end

  # Options passed by the user that may be passed through to the underlying parsing library.
  # @return [Hash]
  attr_reader :other_options

  # Create a new RemoteTable, which is an Enumerable.
  #
  # Options are set at creation using any of the attributes listed... RDoc will say they're "read-only" because you can't set/change them after creation.
  #
  # Does not immediately download/parse... it's lazy-loading.
  #
  # @overload initialize(settings)
  #   @param [Hash] settings Settings including +:url+.
  #
  # @overload initialize(url, settings)
  #   @param [String] url The URL to the local or remote file.
  #   @param [Hash] settings Settings.
  #
  # @example Open an XLSX
  #   RemoteTable.new('http://www.customerreferenceprogram.org/uploads/CRP_RFP_template.xlsx')
  #
  # @example Open a CSV inside a ZIP file
  #   RemoteTable.new 'http://www.epa.gov/climatechange/emissions/downloads10/2010-Inventory-Annex-Tables.zip',
  #                   :filename => 'Annex Tables/Annex 3/Table A-93.csv',
  #                   :skip => 1,
  #                   :pre_select => proc { |row| row['Vehicle Age'].strip =~ /^\d+$/ }
  def initialize(*args)
    @download_count_mutex = ::Mutex.new
    @extend_bang_mutex = ::Mutex.new

    @cache = []
    @download_count = 0

    settings = args.last.is_a?(::Hash) ? args.last.symbolize_keys : {}

    @url = if args.first.is_a? ::String
      args.first
    else
      grab settings, :url
    end
    @format = RemoteTable.guess_format grab(settings, :format)
    if GOOGLE_DOCS_SPREADSHEET.any? { |regex| regex =~ url }
      @url = RemoteTable.google_spreadsheet_csv_url url
      @format = :delimited
    end

    @headers = grab settings, :headers
    if headers.is_a?(::Array) and headers.any?(&:blank?)
      raise ::ArgumentError, "[remote_table] If you specify headers, none of them can be blank"
    end
    @quote_char = grab settings, :quote_char

    @compression = grab(settings, :compression) || RemoteTable.guess_compression(url)
    @packing = grab(settings, :packing) || RemoteTable.guess_packing(url)

    @streaming = grab settings, :streaming
    @warn_on_multiple_downloads = grab settings, :warn_on_multiple_downloads
    @delimiter = grab settings, :delimiter
    @sheet = grab settings, :sheet
    @keep_blank_rows = grab settings, :keep_blank_rows
    @form_data = grab settings, :form_data
    @skip = grab settings, :skip
    @encoding = grab settings, :encoding
    @row_xpath = grab settings, :row_xpath
    @column_xpath = grab settings, :column_xpath
    @row_css = grab settings, :row_css
    @column_css = grab settings, :column_css
    @glob = grab settings, :glob
    @filename = grab settings, :filename
    @cut = grab settings, :cut
    @crop = grab settings, :crop
    @schema = grab settings, :schema
    @schema_name = grab settings, :schema_name
    @pre_select = grab settings, :pre_select
    @pre_reject = grab settings, :pre_reject
    @errata = grab settings, :errata
    @root_node = grab settings, :root_node
    @parser = grab settings, :parser
    @stop_after_untitled_headers = grab settings, :stop_after_untitled_headers

    @other_options = settings

    @local_copy = LocalCopy.new self
    extend!
  end

  # Yield each row.
  #
  # @return [nil]
  #
  # @yield [Hash,Array] A hash or an array depending on whether the RemoteTable has named headers (column names).
  def each
    if fully_cached?
      cache.each do |row|
        yield row
      end
    else
      mark_download!
      preprocess!
      memo = _each do |row|
        parser.call(row).each do |virtual_row|
          virtual_row.row_hash = ::HashDigest.digest3 row
          if errata
            next if errata.rejects? virtual_row
            errata.correct! virtual_row
          end
          next if pre_select and !pre_select.call(virtual_row)
          next if pre_reject and pre_reject.call(virtual_row)
          unless streaming
            cache.push virtual_row
          end
          yield virtual_row
        end
      end
      unless streaming
        fully_cached!
      end
      memo
    end
    nil
  end

  # @deprecated
  alias :each_row :each

  # @return [Array<Hash,Array>] All rows.
  def to_a
    if fully_cached?
      cache.dup
    else
      map { |row| row }
    end
  end

  # @deprecated
  alias :rows :to_a

  # Get a row by row number. Zero-based.
  #
  # @return [Hash,Array]
  def [](row_number)
    if fully_cached?
      cache[row_number]
    else
      to_a[row_number]
    end
  end

  # Clear the row cache in case it helps your GC.
  #
  # @return [nil]
  def free
    @fully_cached = false
    cache.clear
    nil
  end

  private

  def preprocess!
    # noop, overridden sometimes
  end

  def mark_download!
    @download_count_mutex.synchronize do
      @download_count += 1
    end
    if warn_on_multiple_downloads and download_count > 1
      ::Kernel.warn "[remote_table] #{url} has been downloaded #{download_count} times."
    end
  end

  def fully_cached!
    @fully_cached = true
  end

  def fully_cached?
    !!@fully_cached
  end

  def transliterate_to_utf8(str)
    if str.is_a?(::String)
      ::ActiveSupport::Inflector.transliterate str
    end
  end

  def assume_utf8(str)
    if str.is_a?(::String) and ::RUBY_VERSION >= '1.9'
      str.encode! EXTERNAL_ENCODING
    else
      str
    end
  end

  def grab(settings, k)
    user_specified = false
    memo = nil
    if (old_names = OLD_SETTING_NAMES[k]) and old_names.any? { |old_k| settings.has_key?(old_k) }
      user_specified = true
      memo = old_names.map { |old_k| settings.delete(old_k) }.compact.first
    end
    if settings.has_key?(k)
      user_specified = true
      memo = settings.delete k
    end
    if not user_specified and DEFAULT.has_key?(k)
      memo = DEFAULT[k]
    end
    if memo and (valid = VALID[k]) and not valid.include?(memo.to_sym)
      raise ::ArgumentError, %{[remote_table] #{k.inspect} => #{memo.inspect} is not a valid setting. Valid settings are #{valid.inspect}.}
    end
    memo
  end

  def extend!
    return if @extend_bang
    @extend_bang_mutex.synchronize do
      return if @extend_bang
      @extend_bang = true
      format_module = if format
        RemoteTable.const_get format.to_s.camelcase
      elsif format = RemoteTable.guess_format(local_copy.path)
        @format = format
        RemoteTable.const_get format.to_s.camelcase
      else
        Delimited
      end
      extend format_module
      after_extend if respond_to?(:after_extend)
    end
  end
end
